### PR TITLE
Add k-combinations generator

### PR DESCRIPTION
**Pull request by arcondello**
Creates a bqm that is minimized when k of n variables are equal to 1.

Note that the function has not been added to the online documentation, see #433.

### CODE REVIEW

File: dimod/generators/constraints.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+try:
+    import collections.abc as abc
+except ImportError:
+    import collections as abc
+
+from dimod.binary_quadratic_model import BinaryQuadraticModel
+from dimod.decorators import graph_argument, vartype_argument
+from dimod.vartypes import BINARY
+
+__all__ = 'combinations',
```
Comment (randomir) [inline dimod/generators/constraints.py]: I noticed your increasing usage of this *implicit tuple assignment to `__all__`* pattern, and it's killing me. :)

If you hate `__all__` as a list (*the standard* pattern), use tuples - but please make them explicit. :)

Comment (arcondello) [reply dimod/generators/constraints.py]: Well if it's the *standard* fine, but my way is still much prettier

Comment (randomir) [reply dimod/generators/constraints.py]: Well, it's the standard way for a reason: https://zen-of-python.info/explicit-is-better-than-implicit.html#2 :)

Comment (arcondello) [reply dimod/generators/constraints.py]: Feel like I could make a case for readability counts if I had to

Comment (randomir) [reply dimod/generators/constraints.py]: Not sure I follow - are you suggesting that trailing comma is more readable than an explicit tuple or list?

Comment (arcondello) [reply dimod/generators/constraints.py]: To be clear I will go back to the ugly list way! But consider this to be my official desired exception to python standards

Comment (randomir) [reply dimod/generators/constraints.py]: Readability impact:
```
__all__ = ['combinations']
# vs
__all__ = 'combinations',
```

Gotcha impact:
```
__all__ = 'combinations',
          'ignored'

# to work, you need the ugly escape:
__all__ = 'combinations',\
          'ignored'
```

```
# small input difference, major impact
__all__ = 'combinations',
__all__ = 'combinations'
```

Comment (arcondello) [reply dimod/generators/constraints.py]: I was going to argue that my way saves from
```
__all__ = ['hello', 'world']
__all__ = ['hello' 'world']
```
but turns out
```
a = 'hello' 'world'
```
is valid in python, who knew

Comment (arcondello) [reply dimod/generators/constraints.py]: Also the last case would throw an error

Comment (randomir) [reply dimod/generators/constraints.py]: What case would throw an error?

Comment (randomir) [reply dimod/generators/constraints.py]: And regarding the "who knew" part; that's actually a pattern I like to use :)

https://github.com/dwavesystems/dwave-hybrid/blob/master/hybrid/flow.py#L637

Comment (arcondello) [reply dimod/generators/constraints.py]: I meant `__all__ = 'combinations'` or `__all__ = ['hello' 'world']`, assuming `helloworld` was not an object in the module.
But I mis-remembered, it only throws an error if you try to `from example import *` and the `__all__` is example has something not actually present.

**Pull request merged**